### DISCUSSION TITLE
Support MIMO beamforming on gNB

### DISCUSSION THREAD

**Discussion by zenghuacheng** [srsran/srsRAN_Project#76] [General Help] [answered]
Hello:

I wonder if the current code can support transmitter-side beamforming for a two antenna gNB. I do see the receiver has code for 2x2 zero-forcing channel estimation and equalization. Does the transmitter side support 2x2 beamforming? If not, does the transmitter support multiple data streams to USRP? We want to implement a beamforming algorithm by three steps: (1) gNB's RF calibration; (2) uplink 2x2 channel estimation at gNB, (3) downlink beamforming at gNB (using channel reciprocity). Is it doable on srsRAN_Project? We have implemented and tested this solution in real time on USRP devices, but want to implement it on srsRAN_Project to evaluate its performance in 5G systems.

Thanks in advance!

Huacheng Zeng

**Comment by ismagom** [comment, marked as answer]
We are currently working on adding support for MIMO. Beamforming will come after that.

**Comment by frankist** [comment]
DL MIMO 2x2 and 4x4 has been added in the main branch.

**Comment by gctziavas** [comment]
Hello @ismagom, @frankist. 
Have there been any updates on the beamforming? 
And if you have implemented it, do you recommend any RUs supporting it?

Thank you!
